File: client/types/payload/request/employee.request.ts
import { z } from "zod/v3";
import { MarriageStatus } from "@/types/enums/marriage.status";
import { EmployeeStatus } from "@/types/enums/employee.status";
import { Gender } from "@/types/enums/gender";

export const EmployeeRequest = z.object({
  name: z
    .string()
    .min(1, "Name is required.")
    .max(50, "Name must not exceed 50 characters."),

  email: z
    .string()
    .min(1, "Email is required.")
    .email("Email format is invalid.")
    .max(100, "Email must not exceed 100 characters."),

  phoneNumber: z
    .string()
    .min(1, "Phone number is required.")
    .max(20, "Phone number must not exceed 20 characters."),

  taxId: z.string().regex(/^[0-9]{15}$/, "Tax ID must be exactly 15 digits"),

  baseSalary: z
    .number()
    .positive("Base salary must be greater than 0.")
    .refine((v) => Number.isFinite(v), "Base salary format is invalid."),

  marriageStatus: MarriageStatus,

  numberOfDependents: z
    .number()
    .int()
    .min(0, "Number of dependents must be 0 or greater.")
    .optional(),

  employeeStatus: EmployeeStatus,

  birthDate: z.string({ required_error: "Birth date is required" }).refine(
    (val) => {
      const date = new Date(val);
      return !isNaN(date.getTime()) && date <= new Date();
    },
    { message: "Birth date cannot be in the future" },
  ),

  gender: Gender,

  roleIds: z.array(z.number().int()).min(1, "Role must not be empty."),

  departmentId: z.number().int({
    message: "Department ID is required.",
  }),
});

export type EmployeeRequest = z.infer<typeof EmployeeRequest>;

File: backend/src/main/java/com/lerneon/backend/models/payload/request/EmployeeRequest.java
package com.lerneon.backend.models.payload.request;

import com.lerneon.backend.models.annotations.ValidEnum;
import com.lerneon.backend.models.enums.EmployeeStatus;
import com.lerneon.backend.models.enums.Gender;
import com.lerneon.backend.models.enums.MarriageStatus;
import jakarta.validation.constraints.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class EmployeeRequest {

    @NotBlank(message = "Name is required.")
    @Size(max = 50, message = "Name must not exceed 50 characters.")
    private String name;

    @NotBlank(message = "Email is required.")
    @Email(message = "Email format is invalid.")
    @Size(max = 100, message = "Email must not exceed 100 characters.")
    private String email;

    @NotBlank(message = "Phone number is required.")
    @Size(max = 20, message = "Phone number must not exceed 20 characters.")
    private String phoneNumber;

    @NotBlank(message = "Tax ID is required.")
    @Size(max = 20, message = "Tax ID must not exceed 20 characters.")
    private String taxId;

    @NotBlank(message = "Employee Status is required")
    @ValidEnum(enumClass = EmployeeStatus.class, message = "Employee Status must be one of: ACTIVE, RESIGNED")
    private String employeeStatus;

    @NotNull(message = "Base salary is required.")
    @Digits(integer = 17, fraction = 2, message = "Base salary format is invalid.")
    @DecimalMin(value = "0.0", inclusive = false, message = "Base salary must be greater than 0.")
    private BigDecimal baseSalary;

    @NotBlank(message = "Marriage Status is required")
    @ValidEnum(enumClass = MarriageStatus.class, message = "Marriage Status must be one of: MARRIED, SINGLE, DIVORCED")
    private String marriageStatus;

    @Min(value = 0, message = "Number of dependents must be 0 or greater.")
    private Integer numberOfDependents;

    @NotNull(message = "Established date is required")
    @PastOrPresent(message = "Established date cannot be in the future")
    private LocalDate birthDate;

    @NotBlank(message = "Gender is required")
    @ValidEnum(enumClass = Gender.class, message = "Gender must be one of: MAN, WOMAN, UNKNOWN")
    private String gender;

    @NotEmpty(message = "Role IDs must not be empty.")
    private List<Integer> roleIds;

    @NotNull(message = "Department ID is required.")
    private Integer departmentId;
}
